fullrt/dht.go

```diff
@@ -834,35 +834,41 @@ func (dht *FullRT) execOnMany(ctx context.Context, fn func(context.Context, peer
 	putctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	waitAllCh := make(chan struct{}, len(peers))
+	waitFailCh := make(chan struct{})
+	waitSuccessCh := make(chan struct{})
 	numSuccessfulToWaitFor := int(float64(len(peers)) * dht.waitFrac)
-	waitSuccessCh := make(chan struct{}, numSuccessfulToWaitFor)
 	for _, p := range peers {
 		go func(p peer.ID) {
 			fnCtx, fnCancel := context.WithTimeout(putctx, dht.timeoutPerOp)
 			defer fnCancel()
 			err := fn(fnCtx, p)
 			if err != nil {
 				logger.Debug(err)
+				waitFailCh <- struct{}{}
 			} else {
 				waitSuccessCh <- struct{}{}
 			}
-			waitAllCh <- struct{}{}
 		}(p)
 	}
 
-	numSuccess, numDone := 0, 0
-	t := time.NewTimer(time.Hour)
-	for numDone != len(peers) {
+	var numFail, numSuccess int
+	t := time.NewTimer(0)
+	if !t.Stop() {
+		<-t.C
+	}
+	defer t.Stop()
+	for numFail+numSuccess != len(peers) {
 		select {
-		case <-waitAllCh:
-			numDone++
+		case <-waitFailCh:
+			numFail++
 		case <-waitSuccessCh:
 			if numSuccess >= numSuccessfulToWaitFor {
+				if !t.Stop() {
+					<-t.C
```

Comment:
this will block the first time it's called because the timer is already stopped which means `t.Stop()` is false, and we've already emptied out the channel.

Doing it this way we need some boolean to check for the first time we've reset the timer. This follows roughly the same pattern as the pause detection timer in this function https://github.com/ipfs/go-ipfs-provider/blob/d391dae4a595473f6797eb5d5b803a529a7bbdbc/batched/system.go#L130

Reply:
nm, looks like we're discussing this in #719